systems/controllers/cp_traj_gen.cc

```diff
@@ -0,0 +1,397 @@
+  is_quaternion_ = multibody::CheckFloatingBase(tree);
+}
+
+
+EventStatus CPTrajGenerator::DiscreteVariableUpdate(
+  const Context<double>& context,
+  DiscreteValues<double>* discrete_state) const {
+
+  // Read in finite state machine
+  const BasicVector<double>* fsm_output = (BasicVector<double>*)
+      this->EvalVectorInput(context, fsm_port_);
+  VectorXd fsm_state = fsm_output->get_value();
+
+  auto prev_fsm_state = discrete_state->get_mutable_vector(
+                          prev_fsm_state_idx_).get_mutable_value();
+
+  if (fsm_state(0) != prev_fsm_state(0)) { //if at touchdown
+    prev_fsm_state(0) = fsm_state(0);
+
+    auto swing_foot_pos_td = discrete_state->get_mutable_vector(
+                               prev_td_swing_foot_idx_).get_mutable_value();
+    auto prev_td_time = discrete_state->get_mutable_vector(
+                          prev_td_time_idx_).get_mutable_value();
+
+    // Read in current state
+    const OutputVector<double>* robot_output = (OutputVector<double>*)
+        this->EvalVectorInput(context, state_port_);
+
+    // Get time
+    double timestamp = robot_output->get_timestamp();
+    double current_time = static_cast<double>(timestamp);
+    prev_td_time(0) = current_time;
+
+    // Kinematics cache and indices
+    KinematicsCache<double> cache = tree_->CreateKinematicsCache();
+    VectorXd q = robot_output->GetPositions();
+    // Modify the quaternion in the begining when the state is not received from
+    // the robot yet (cannot have 0-norm quaternion when using doKinematics)
+    if (is_quaternion_) {
+      q.segment(3, 4) = NormalizeQuaternion(q.segment(3, 4));
+    }
+    cache.initialize(q);
+    tree_->doKinematics(cache);
+    int swing_foot_idx = (fsm_state(0) == right_stance_) ?
+                           left_foot_idx_ : right_foot_idx_;
+    Vector3d pt_on_swing_foot = (fsm_state(0) == right_stance_) ?
+                                pt_on_left_foot_ : pt_on_right_foot_;
+
+    // Swing foot position (Forward Kinematics) and velocity at touchdown
+    Eigen::Isometry3d swing_foot_body_pose =
+        tree_->CalcBodyPoseInWorldFrame(cache, tree_->get_body(swing_foot_idx));
+    Vector3d swing_foot_body_pos = swing_foot_body_pose.translation();
+    Eigen::MatrixXd swing_foot_body_rot = swing_foot_body_pose.linear();
+    swing_foot_pos_td = swing_foot_body_pos +
+                        swing_foot_body_rot * pt_on_swing_foot;
+  }
+
+  return EventStatus::Succeeded();
+}
+
+
+Vector2d CPTrajGenerator::calculateCapturePoint(const Context<double>& context,
+    const OutputVector<double>* robot_output,
+    const double end_time_of_this_interval) const {
+  // Read in finite state machine
+  const BasicVector<double>* fsm_output = (BasicVector<double>*)
+      this->EvalVectorInput(context, fsm_port_);
+  VectorXd fsm_state = fsm_output->get_value();
+
+  // Get stance foot position and index
+  KinematicsCache<double> cache = tree_->CreateKinematicsCache();
```

Comment:
This comment goes for this file and others--you might be paying a significant computational penalty because of the way you split up these methods _without_ leveraging caching. Many of your methods recompute the same `KinematicsCache`. `MBP` will help with this, but you also might be better off by passing the cache as inputs to many methods.

Reply:
What is the best way of doing the caching? I'm thinking create a class `CentralKinematicsCache` with two members
- `kinematics_cache`
- `time` which stores the time when the kinematics cache is updated.

 and two functions
- `initialize()` which initiailizes the members
- `update()` which updates kinematics cache if the current time is greater than `time`, and return a const reference to `kinematics_cache`.

I construct `CentralKinematicsCache` in run_osc_walking_controller.cc and pass it into the leaf systems that need it?